docs/configuration/indexes.md

```diff
@@ -0,0 +1,114 @@
+# Package indexes
+
+By default, uv uses the [Python Package Index (PyPI)](https://pypi.org) for dependency resolution
+and package installation. However, uv can be configured to use other package indexes, including
+private indexes, via the `[[tool.uv.index]]` configuration option (and `--index`, its analogous
+command-line option).
+
+## Defining an index
+
+To include an additional index when resolving dependencies, add a `[[tool.uv.index]]` entry to your
+`pyproject.toml`:
+
+```toml
+[[tool.uv.index]]
+# Optional, explicit name for the index.
+name = "pytorch"
+# Required URL for the index. Expects a repository compliant with PEP 503 (the simple repository API).
+url = "https://download.pytorch.org/whl/cpu"
+```
+
+Indexes are prioritized in the order in which they’re defined, such that the first index listed in
+the configuration file is the first index consulted when resolving dependencies, with indexes
+provided via the command line taking precedence over those in the configuration file.
+
+By default, uv includes the Python Package Index (PyPI) as the "default" index, i.e., the index used
+when a package is not found on any other index. To exclude PyPI from the list of indexes, set
+`default = true` on another index entry (or use the `--default-index` command-line option):
+
+```toml
+[[tool.uv.index]]
+name = "pytorch"
+url = "https://download.pytorch.org/whl/cpu"
+default = true
+```
+
+The default index is always treated as lowest priority, regardless of its position in the list of
+indexes.
+
+## Pinning a package to an index
+
+A package can be pinned to a specific index by specifying the index in its `tool.uv.sources` entry.
+For example, to ensure that `torch` is _always_ installed from the `pytorch` index, add the
+following to your `pyproject.toml`:
+
+```toml
+[tool.uv.sources]
+torch = { index = "pytorch" }
+
+[[tool.uv.index]]
+name = "pytorch"
+url = "https://download.pytorch.org/whl/cpu"
+```
+
+An index can be marked as `explicit = true` to prevent packages from being installed from that index
+unless explicitly pinned to it. For example, to ensure that `torch` is _only_ installed from the
+`pytorch` index, add the following to your `pyproject.toml`:
+
+```toml
+[tool.uv.sources]
+torch = { index = "pytorch" }
+
+[[tool.uv.index]]
+name = "pytorch"
+url = "https://download.pytorch.org/whl/cpu"
+explicit = true
+```
+
+Named indexes referenced via `tool.uv.sources` must be defined within the project's `pyproject.toml`
+file; indexes provided via the command-line, environment variables, or user-level configuration will
+not be recognized.
```

Comment:
As far as I can tell, we silently ignore if we receive a command-line argument that interacts with a named index right?

Separately, why this limitation? What happened to being able to override a named index on a single invocation?

Reply:
> Separately, why this limitation? What happened to being able to override a named index on a single invocation?

It's an implementation detail... It probably should work, but it's hard because those get merged into settings, and we don't look in the settings when resolving explicit, named indexes (we just look in the current workspace).

Reply:
I'm working on this, but it will be a separate PR.

Reply:
I'm considering tracking the "source" of every index (CLI, configuration file, etc.), as field on the index type.